### PR TITLE
Run VSphere tests in self-hosted, refactor test suite into spec per provider, support label filtering, get all tests green

### PR DESCRIPTION
**Pull request by squizzi**
This PR refactors our existing e2e test spec into a more conventional e2e test run.  We now have a `BeforeSuite` that runs and installs/validates the controller.  Prior to this we had all of our e2e tests under a single `Controller` spec, but this breaks it into:

* Controller (`Label(controller)`)
* Providers (`Label(provider)`)
    * AWS (`Label(provider:aws, provider:cloud)`) - `test/e2e/provider_aws_test.go`
    * Azure (`Label(provider:azure, provider:cloud)`) - `test/e2e/provider_azure_test.go`
    * Vsphere (`Label(provider:vsphere, provider:onprem)`) - `test/e2e/provider_vsphere_test.go`
 
The labels for these tests have been documented here: https://github.com/Mirantis/hmc/pull/360/files#diff-3a6cab76ecce19612a704fe0ef89dc83f1ed2df44f7bfa28d47d2529a7949fafR139

Each of the umbrella labels correspond to a GitHub job that is dependent on the `Build and Unit Test` job which provides setup.   The `Controller` tests always run, no matter the label, right now the `Controller` test is disabled because there aren't any tests other than Before/AfterSuite, but we can enable it because that sort of is a test of the controller in itself.

The provider tests are still attached to the `test e2e` label, but I think we can break these out further with `provider` specific GitHub labels at some point.

`provider:onprem` tests run on a self-hosted GitHub runner that has access to Mirantis' network, this is to support providers that do not necessarily have cloud infrastructure we can get to from a GitHub hosted runner -- like VSphere.

This PR also adds support for creating `ClusterIdentity`, `Credential` and their associated `Secret` resources  with a new `clusteridentity` package.

And of course, with testing comes bug fixing, so this PR fixes several different issues discovered along the way.

Closes: #210, Closes: #323

---

Note: I've opened two other issues to further simplify these tests, we could make these a lot easier to write in the future with:
* https://github.com/Mirantis/hmc/issues/427
* https://github.com/Mirantis/hmc/issues/426

### CODE REVIEW

**Comment by squizzi** [comment]
I'd like to break the controller tests into their own reusable workflow with Build/Unit Test/Setup being what runs first always, then move Provider tests into their own thing as well.  Might not be within the scope of this, but leaving this thought here to maybe add a follow-up issue around this.

**Comment by squizzi** [comment]
Note: This run shows a vsphere green: https://github.com/Mirantis/hmc/actions/runs/11153080327/job/31036965900?pr=360

We might see some failures in upcoming runs as Mirantis IT is doing some work on our self-hosted runner.

File: .github/workflows/build_test.yml
Comment (kylewuolle) [inline .github/workflows/build_test.yml:143]: Should this be in here still?

**Review by kylewuolle — CHANGES_REQUESTED** [review]
Looks good just wonder about leaving tmate in there

File: .github/workflows/build_test.yml
Comment (squizzi) [reply .github/workflows/build_test.yml:143]: I'm still using it, I'm going to remove it once I get Azure green

**Review by kylewuolle — DISMISSED** [review]

**Comment by squizzi** [comment]
csi-drivers are going to be the death of me 😂

**Comment by squizzi** [comment]
The `azure-hosted-cp` `ManagedCluster` appears to have hit some sort of bug in the last run: `azuremachines` is reporting a 100% ready status while `machines` is not, and is reporting a `NodeHealthy: false` status.  The `cluster-api-operator` is reporting that the infra is in `status.ready=true` state:

```
I1009 18:24:48.584810       1 machine_controller_phases.go:294] "Infrastructure provider has completed machine infrastructure provisioning and reports status.ready" controller="machine" controllerGroup="cluster.x-k8s.io" controllerKind="Machine" Machine="hmc-system/ci-96157-azure-hosted-md-7fvv6-j8hd5" namespace="hmc-system" name="ci-96157-azure-hosted-md-7fvv6-j8hd5" reconcileID="ddd599f6-0410-4f20-91ae-adf60a9e115f" MachineSet="hmc-system/ci-96157-azure-hosted-md-7fvv6" MachineDeployment="hmc-system/ci-96157-azure-hosted-md" Cluster="hmc-system/ci-96157-azure-hosted" AzureMachine="hmc-system/ci-96157-azure-hosted-md-7fvv6-j8hd5"
I1009 18:24:48.723231       1 machine_controller_phases.go:294] "Infrastructure provider has completed machine infrastructure provisioning and reports status.ready" controller="machine" controllerGroup="cluster.x-k8s.io" controllerKind="Machine" Machine="hmc-system/ci-96157-azure-hosted-md-7fvv6-j8hd5" namespace="hmc-system" name="ci-96157-azure-hosted-md-7fvv6-j8hd5" reconcileID="c7a6b3e7-aef9-4f2a-9d8f-070ad50aaa57" MachineSet="hmc-system/ci-96157-azure-hosted-md-7fvv6" MachineDeployment="hmc-system/ci-96157-azure-hosted-md" Cluster="hmc-system/ci-96157-azure-hosted" AzureMachine="hmc-system/ci-96157-azure-hosted-md-7fvv6-j8hd5"
```

**Review by kylewuolle — DISMISSED** [review]

**Review by slysunkin — DISMISSED** [review]

**Review by slysunkin — DISMISSED** [review]

**Review by kylewuolle — APPROVED** [review]

File: test/e2e/managedcluster/common.go
Comment (a13x5) [inline test/e2e/managedcluster/common.go:55]: Shouldn't be used for vSphere, so here I suggest just return

File: test/e2e/provider_aws_test.go
Comment (a13x5) [inline test/e2e/provider_aws_test.go:104]: I'm suggesting to create actual tasks and reference it here, so it will be _actually_ done sometime in the future

File: test/e2e/provider_vsphere_test.go
Comment (a13x5) [inline test/e2e/provider_vsphere_test.go:64]: The GH issue would be nice for that as well

File: .github/workflows/build_test.yml
Comment (a13x5) [inline .github/workflows/build_test.yml:3]: I also suggest running e2e tests when tags are pushed (`on: push: tags "*"` ), so the release could be validated.

**Review by a13x5 — DISMISSED** [review]
Some minor notes here and there.

I don't have major concerns regarding actual tests, apart from the fact that standalone and hosted CP tests are squashed together. 

I understand the reason for that, since we're reusing standalone cluster and it must be created anyway for hosted CP tests, but I see the following cons of this approach:

- standalone and hosted tests can't be run separately
- If hosted CP fails (and it will a lot I'm sure) it will fail the whole AWS test, which makes it slightly harder to interpret the test.

I know that it's e2e tests and we should run the whole suite, but I believe it makes sense to run only on the release or ad-hoc. Most changes could be tested by simply running tests on standalone cluster.

I will not block this PR because of that, but I a task for the future us would be nice.

File: .github/workflows/build_test.yml
Comment (squizzi) [reply .github/workflows/build_test.yml:11]: This predated this PR so I left it in the new file.

File: Makefile
Comment (squizzi) [reply Makefile:305]: To help support the `NO_CLEANUP` use case and make the tests work if things already exist without failure.  Some of the template specs are immutable, iirc `--force` replaces the spec instead of a standard apply.

File: test/e2e/provider_aws_test.go
Comment (squizzi) [reply test/e2e/provider_aws_test.go:104]: Ya that's fair, I'll create them.

File: .github/workflows/build_test.yml
Comment (squizzi) [reply .github/workflows/build_test.yml:3]: I'm assuming `on.push` just disregards `github.event.pull_request.labels` (hopefully)

File: Makefile
Comment (a13x5) [reply Makefile:305]: Ok, got it. Let's leave it here

File: test/e2e/e2e_suite_test.go
Comment (a13x5) [inline test/e2e/e2e_suite_test.go:126]: Since we can  now have more than one item here we should iterate.

**Review by a13x5 — COMMENTED** [review]
Minor notes

File: test/e2e/e2e_suite_test.go
Comment (squizzi) [reply test/e2e/e2e_suite_test.go:126]: Missed that, thanks

**Review by slysunkin — DISMISSED** [review]

**Review by a13x5 — DISMISSED** [review]
LGTM
PR needs rebasing

**Comment by squizzi** [comment]
@a13x5 rebased

**Review by a13x5 — APPROVED** [review]

**Review by slysunkin — APPROVED** [review]

**Comment by DinaBelova** [comment]
Will be merging w/ admin rights to bypass the checks

Job will not run on this patch  because of `pull_request_target` being the workflow, it needs the actions workflow to live in the base branch before it gets triggered, but this PR adds the actions.

@squizzi will make a follow-up after this PR merge to ensure no issues are present.

**Pull request merged**
